first attempt at the Auction emulator

venkatacrc opened the pull request.

venkatacrc committed:
first attempt at the Auction emulator

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py:
```diff
@@ -1,15 +1,70 @@
     def close(self):
-        pass
\ No newline at end of file
+        self.bids = [] 
```
missing newline at end of file

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/main.py:
```diff
@@ -0,0 +1,14 @@
+
+flags.DEFINE_integer('market_price', 10, 'mkt price index inside the bid reqs')
+flags.DEFINE_integer('click', 11, 'click index inside the bid reqs')
+flags.DEFINE_integer('conversion', 12, 'conversion index inside the bid reqs')
```
missing new line at end of file.

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/main.py:
```diff
@@ -0,0 +1,14 @@
+
+
+import gflags
```
Do we really need to use `gflags` here? What benefits will it bring down the line?

I think that for the time being we can get rid of `main.py` and replace it with a config file. We should probably start with tools from the python standard library until we have a good reason to move away from them.

I'd create a file named `config.cfg` with the following structure:
```
[bid_request]
  market_price = 10
  click = 11
  conversion = 12
```

then you can create a hidden function on your class to load this config file and call it when you initialize your agent.

For example:

```
    def _load_config(self):
        cfg = configparser.ConfigParser()
        cfg.read('./config.cfg')
        self.bid_request_fields = {}
        for item in cfg['bid_request']:
            self.bid_request_fields[item] = int(cfg['bid_request'][item])

```

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py:
```diff
@@ -1,15 +1,70 @@
+        fin.close()
+
+
+    def _get_bid_fields(self):
```
This function should probably return a dictionary with the field name and the value instead of using predetermined variable names. That way we can easily change what gets passed to the user by changing `config.cfg` as I suggested in the `main.py` file.

What is the benefit of having `mkt_price`, `click` and `conversion` stored at the class level?

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py:
```diff
@@ -1,15 +1,70 @@
+            bid_vals = bid.split(" ")
+            self.bids.append(bid_vals) 
+        self.num_episodes = len(self.bids)
+        fin.close()
```
A more pythonic way of writing this function would be:

```
    with open(file_in, 'r') as f:
        self.bid_requests = [br.rstrip('\n').split('\t') for br in f]
```

We should probably also think about a nice way to deal with very large datasets. Our training data will probably not fit in memory, so we could load only one (or two) day(s) at a time.

bguisard commented on src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py:
```diff
@@ -1,15 +1,70 @@
+        Reset the OpenAI Gym Auction Emulator environment.
+        """
+        self._step = 0
+        bid_request = self.bids[self._step]
```
we don't need to assign this value to a new variable, we can read directly from `self.bids[self._step]` when needed.

bguisard commented on src/gym-auction_emulator/setup.py:
```diff
@@ -3,5 +3,5 @@
     name='gym_auction_emulator',
     version='0.0.1',
-    install_requires=['gym']
+    install_requires=['gym', 'gflags']
```
missing new line at end of file.

bguisard reviewed (commented):
Thanks for this first PR, this is a good first step.

I have left several comments that I think will simplify this first version.

Most notably what is missing from this first step is a toy dataset (we can take the first 100 rows of the iPinYou dataset for example) and a very simple agent that interacts with this environment, let's say the only thing it does is constantly bid $1 for every bid request it receives. That way we can at least test if things are working as expected.

Also I think now it would be a great time to clean up your repo and replace the placeholders on the template repository with the pieces we will use. This should be done in a separate branch and most likely doesn't need a PR.

venkatacrc commented:
Thank you for the detailed feedback Bruno! It will help me a lot reiterate
my next version.

On Thu, Sep 19, 2019 at 8:47 PM Bruno Guisard <notifications@github.com>
wrote:

> *@bguisard* commented on this pull request.
>
> Thanks for this first PR, this is a good first step.
>
> I have left several comments that I think will simplify this first version.
>
> Most notably what is missing from this first step is a toy dataset (we can
> take the first 100 rows of the iPinYou dataset for example) and a very
> simple agent that interacts with this environment, let's say the only thing
> it does is constantly bid $1 for every bid request it receives. That way we
> can at least test if things are working as expected.
>
> Also I think now it would be a great time to clean up your repo and
> replace the placeholders on the template repository with the pieces we will
> use. This should be done in a separate branch and most likely doesn't need
> a PR.
> ------------------------------
>
> In
> src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326453895>
> :
>
> > \ No newline at end of file
> +        self.bids = []
>
> missing newline at end of file
> ------------------------------
>
> In src/gym-auction_emulator/gym_auction_emulator/envs/main.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326453921>
> :
>
> > @@ -0,0 +1,14 @@
> +
> +
> +import gflags
> +
> +"""
> +Dataset column indexes
> +TODO: replace dummy indexes with real indexes
> +"""
> +
> +FLAGS = gflags.FLAGS
> +
> +flags.DEFINE_integer('market_price', 10, 'mkt price index inside the bid reqs')
> +flags.DEFINE_integer('click', 11, 'click index inside the bid reqs')
> +flags.DEFINE_integer('conversion', 12, 'conversion index inside the bid reqs')
>
> missing new line at end of file.
> ------------------------------
>
> In src/gym-auction_emulator/gym_auction_emulator/envs/main.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326457770>
> :
>
> > @@ -0,0 +1,14 @@
> +
> +
> +import gflags
>
> Do we really need to use gflags here? What benefits will it bring down
> the line?
>
> I think that for the time being we can get rid of main.py and replace it
> with a config file. We should probably start with tools from the python
> standard library until we have a good reason to move away from them.
>
> I'd create a file named config.cfg with the following structure:
>
> [bid_request]
>   market_price = 10
>   click = 11
>   conversion = 12
>
> then you can create a hidden function on your class to load this config
> file and call it when you initialize your agent.
>
> For example:
>
>     def _load_config(self):
>         cfg = configparser.ConfigParser()
>         cfg.read('./config.cfg')
>         self.bid_request_fields = {}
>         for item in cfg['bid_request']:
>             self.bid_request_fields[item] = int(cfg['bid_request'][item])
>
>
> ------------------------------
>
> In
> src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326457951>
> :
>
> > +        """
> +        Args:
> +            file_in: Dataset input file
> +        Populates the bid requests to self.bids list.
> +        """
> +        fin = open(file_in, 'r')
> +        self._step = 0
> +        self.bids = []
> +        for bid in fin:
> +            bid_vals = bid.split(" ")
> +            self.bids.append(bid_vals)
> +        self.num_episodes = len(self.bids)
> +        fin.close()
> +
> +
> +    def _get_bid_fields(self):
>
> This function should probably return a dictionary with the field name and
> the value instead of using predetermined variable names. That way we can
> easily change what gets passed to the user by changing config.cfg as I
> suggested in the main.py file.
>
> What is the benefit of having mkt_price, click and conversion stored at
> the class level?
> ------------------------------
>
> In
> src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326459728>
> :
>
> > +    the bid requests reading the dataset files.
> +    """
> +    def __init__(self, file_in):
> +        """
> +        Args:
> +            file_in: Dataset input file
> +        Populates the bid requests to self.bids list.
> +        """
> +        fin = open(file_in, 'r')
> +        self._step = 0
> +        self.bids = []
> +        for bid in fin:
> +            bid_vals = bid.split(" ")
> +            self.bids.append(bid_vals)
> +        self.num_episodes = len(self.bids)
> +        fin.close()
>
> A more pythonic way of writing this function would be:
>
>     with open(file_in, 'r') as f:
>         self.bid_requests = [br.rstrip('\n').split('\t') for br in f]
>
> We should probably also think about a nice way to deal with very large
> datasets. Our training data will probably not fit in memory, so we could
> load only one (or two) day(s) at a time.
> ------------------------------
>
> In
> src/gym-auction_emulator/gym_auction_emulator/envs/auction_emulator_env.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326461052>
> :
>
> >      def reset(self):
> -        pass
> +        """
> +        Reset the OpenAI Gym Auction Emulator environment.
> +        """
> +        self._step = 0
> +        bid_request = self.bids[self._step]
>
> we don't need to assign this value to a new variable, we can read directly
> from self.bids[self._step] when needed.
> ------------------------------
>
> In src/gym-auction_emulator/setup.py
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1#discussion_r326461143>
> :
>
> > @@ -3,5 +3,5 @@
>  setup(
>      name='gym_auction_emulator',
>      version='0.0.1',
> -    install_requires=['gym']
> +    install_requires=['gym', 'gflags']
>
> missing new line at end of file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/venkatacrc/Budget_Constrained_Bidding/pull/1?email_source=notifications&email_token=ACLTR5O3OAYRDIGIJMXQ3R3QKRBUJA5CNFSM4IYQIOTKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFLUOUY#pullrequestreview-290932563>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACLTR5IRNOTBQ6RZ54Z4LNDQKRBUJANCNFSM4IYQIOTA>
> .
>

venkatacrc committed:
applying bruno's comments

bguisard commented:
I've removed myself from the reviewers. Can you add me again once this is ready for another round of review?

venkatacrc committed:
add toy dataset + toy constant bidding agent example

venkatacrc committed:
take care of eof

venkatacrc committed:
take care of eof
